beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -441,6 +441,47 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             .map(|result| result.map_err(|e| e.into())))
     }
 
+    /// Iterate through the current chain to find the slot intersecting with the given beacon state.
+    /// The maximum depth this will search is `SLOTS_PER_HISTORICAL_ROOT`, and if that depth is reached
+    /// and no intersection is found, the finalized slot will be returned.
+    pub fn find_reorg_slot(
+        &self,
+        new_state: &BeaconState<T::EthSpec>,
+        new_block_root: &Hash256,
+    ) -> Result<Slot, Error> {
+        self.with_head(|snapshot| {
+            let slot = snapshot.beacon_block.slot();
+
+            // Check the head block
+            if &snapshot.beacon_block_root == new_block_root {
+                return Ok(slot);
+            }
+
+            let next_slot = slot - Slot::new(1);
```

Comment:
It might be nice to explicitly use `saturating_sub` or add a comment here to show that we've considered underflow.